mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp

```diff
@@ -1332,6 +1332,120 @@ struct TosaFoldConstantMaximum
+        AccumulatorType acc = static_cast<AccumulatorType>(0);
+        auto indexA =
+            indexToOffset(inputAShape, {batch, height, /*channel=*/0});
+        auto indexB = indexToOffset(inputBShape, {batch, /*channel=*/0, width});
+        for (int64_t channel = 0; channel < channelSize;
+             ++channel, ++indexA, indexB += widthSize) {
+          acc += (convertToAccType(matrixAVals[indexA]) *
+                  convertToAccType(matrixBVals[indexB]));
+        }
+        outputValues[indexOut] = acc;
+      }
+    }
+  }
+
+  return outputValues;
+}
+
+DenseElementsAttr matmul(ShapedType outputType, ElementsAttr matrixA,
+                         ElementsAttr matrixB) {
+
+  auto baseType = outputType.getElementType();
+  if (auto intType = dyn_cast<IntegerType>(outputType.getElementType())) {
+    // For integer types, accumulate values in int64_t to allow support for i8
+    // and i16 that accumulates with i32 and i48, respectively.
+    auto values = matmul<int64_t, APInt>(
+        outputType, matrixA, matrixB,
+        [&](const APInt &val) { return val.getSExtValue(); });
+
+    // Correct int32_t to the correct output type.
+    std::vector<APInt> apintValues;
+    llvm::transform(values, std::back_inserter(apintValues),
+                    [&](const int64_t &val) {
+                      APInt apIntVal(baseType.getIntOrFloatBitWidth(), val);
+                      return apIntVal;
+                    });
+    return DenseElementsAttr::get(outputType, apintValues);
+  }
+
+  assert(isa<FloatType>(baseType) && "Unknown element type.");
+  FloatType fpType = cast<FloatType>(baseType);
+
+  // For FP types, accumulate values in float to cover all cases for FP
+  // matmul. This is safe since tosa supports at most f32 type.
+  auto values = matmul<float, APFloat>(
+      outputType, matrixA, matrixB,
+      [&](const APFloat &val) { return val.convertToFloat(); });
+
+  // Convert float values to the correct output type.
+  std::vector<APFloat> apfloatValues;
+  llvm::transform(values, std::back_inserter(apfloatValues),
+                  [&](float val) -> llvm::APFloat {
+                    bool ignored;
+                    APFloat apFloat(val);
+                    apFloat.convert(fpType.getFloatSemantics(),
+                                    llvm::APFloat::rmNearestTiesToEven,
+                                    &ignored);
+                    return apFloat;
+                  });
+  return DenseElementsAttr::get(outputType, apfloatValues);
+}
+
+struct TosaFoldConstantMatmul : public TosaFoldConstantBase<tosa::MatMulOp> {
+  using TosaFoldConstantBase::TosaFoldConstantBase;
+
+  LogicalResult matchAndRewrite(tosa::MatMulOp op,
+                                PatternRewriter &rewriter) const override {
+    auto outputType = cast<ShapedType>(op.getType());
+    // TOSA doesn't support quantized types.
+    if (!outputType.getElementType().isIntOrIndexOrFloat())
+      return failure();
+
```

Comment:
It seems [zero points for A and B](https://www.mlplatform.org/tosa/tosa_spec.html#_matmul) could also be given and the implementation doesn't take them into account for now. Maybe just not fold if there are zero points and they are non-zero?

Reply:
Implemented and test added.